library/src/main/java/com/novoda/downloadmanager/MigrationFactory.java

```diff
@@ -35,14 +35,19 @@ static Migrator createVersionOneToVersionTwoMigrator(Context context, File datab
             return Migrator.NO_OP;
         }
 
-        SQLiteDatabase sqLiteDatabase = SQLiteDatabase.openDatabase(databasePath.getAbsolutePath(), null, 0);
-        SqlDatabaseWrapper database = new SqlDatabaseWrapper(sqLiteDatabase);
-
-        MigrationExtractor migrationExtractor = new MigrationExtractor(database);
-        RoomDownloadsPersistence downloadsPersistence = RoomDownloadsPersistence.newInstance(context);
-        InternalFilePersistence internalFilePersistence = new InternalFilePersistence();
-        internalFilePersistence.initialiseWith(context);
-        return new VersionOneToVersionTwoMigrator(migrationExtractor, downloadsPersistence, internalFilePersistence, database, migrationCallback);
+        if (INSTANCE == null) {
```

Comment:
I figured it was better to have the one return and just modify the `INSTANCE` rather than having two 🤔  now I am unsure 😂

Reply:
always struggle with multiple return points vs nesting as well. if you considered that it's fine. happy to merge or wait for another review.

Reply:
I'd like @joetimmins or @zegnus to take a look too @takecare. Thanks for reviewing 😃

Reply:
_nudge_

Reply:
I think what you've done with the `INSTANCE` here is fine but it could help readability to pull out another method for lines 39-45. So you end up with something like 
```
        if (INSTANCE == null) {
            INSTANCE = buildMigrator(context, databasePath, migrationCallback);
        }
        return INSTANCE;
```

Reply:
Or even 
```
return INSTANCE == null ? createInstance(context, databasePath, migrationCallback) : INSTANCE;
```

Reply:
Because the singleton is the executor, this is no longer needed

Reply:
@zegnus if you don't have this then you are always stuck with the `V1Migrator` even when it is not being used, it's allocating more memory than the `no-op` which is why I kept this.

Reply:
Yes, you are right @zegnus. We use the executor and the `ServiceBinder` to drive the single instance nature.